src/sql/finalize.sql

```diff
@@ -154,6 +154,87 @@ IMMUTABLE STRICT PARALLEL SAFE LANGUAGE c AS 'MODULE_PATHNAME', '_vchordg_amhand
 CREATE FUNCTION vchordg_prewarm(regclass) RETURNS TEXT
 STRICT LANGUAGE c AS 'MODULE_PATHNAME', '_vchordg_prewarm_wrapper';
 
+CREATE FUNCTION vchordrq_evaluate_query_recall(
+    query_text text,
+    exact_search boolean default False,
+    gt_epsilon real default 1.9
+)
+RETURNS real
+LANGUAGE plpgsql
+STRICT
+AS $$
+DECLARE
+    query_results tid[];
+    gt_results tid[];
+    match_count integer := 0;
+    gt_count integer;
+    recall real;
+    iter_ctid tid;
+    top_k integer;
+    query_probes text;
+    query_epsilon text;
+    gt_probes text;
+
+BEGIN
+    IF NOT exact_search THEN
+        BEGIN
+            query_probes := current_setting('vchordrq.probes');
+            query_epsilon := current_setting('vchordrq.epsilon');
+        EXCEPTION WHEN OTHERS THEN
+            RAISE EXCEPTION 'Could not read vchordrq GUCs. Is vchord shared library loaded? Error: %', SQLERRM;
+        END;
+    END IF;
+
+    BEGIN
+        FOR iter_ctid IN EXECUTE query_text LOOP
+            query_results := array_append(query_results, iter_ctid);
+        END LOOP;
+    EXCEPTION WHEN OTHERS THEN
+        RAISE EXCEPTION 'Error executing ANN query "%": %', query_text, SQLERRM;
+    END;
+
+    top_k := array_length(query_results, 1);
+    IF top_k IS NULL OR top_k = 0 THEN
+        RAISE WARNING 'ANN query returned 0 results. Please check and run the query_text.';
+        RETURN 0.0;
+    END IF;
+
+    BEGIN
+        IF exact_search THEN
+            SET LOCAL vchordrq.enable = off;
+        ELSE
+            IF query_probes = '' THEN
+                gt_probes := '';
+            ELSIF position(',' in query_probes) > 0 THEN
+                gt_probes := '9999,9999';
+            ELSE
+                gt_probes := '9999';
+            END IF;
+            EXECUTE format('SET LOCAL "vchordrq.probes" = %L', gt_probes);
+            EXECUTE format('SET LOCAL "vchordrq.epsilon" = %L', gt_epsilon);
```

Comment:
What if `vchordrq.max_scan_tuples`, `vchordrq.maxsim_refine`, `vchordrq.maxsim_threshold` are used by users?

I still suspect that we shouldn't be using the index as ground truth at all.